transaction/std/src/transaction_builder.rs

```diff
@@ -82,7 +140,78 @@ impl<FPR: FogPubkeyResolver> TransactionBuilder<FPR> {
         recipient: &PublicAddress,
         rng: &mut RNG,
     ) -> Result<(TxOut, TxOutConfirmationNumber), TxBuilderError> {
-        self.add_output_with_fog_hint_address(value, recipient, recipient, rng)
+        let (mut tx_out, conf_num) = self.add_output_with_fog_hint_address(
+            value,
+            recipient,
+            recipient,
+            MemoPayload::default(),
+            rng,
+        )?;
+        // Hack: we can't build authenticated sender memo until we know TxOut public
+        // key, so if we need to do that, modify the TxOut.e_memo now after it
+        // has been added to outputs_and_shared_secrets
+        if let Some(cred) = &self.sender_memo_credential {
+            let (ref mut tx_out_ref, ref shared_secret) = self
+                .outputs_and_shared_secrets
+                .last_mut()
+                .expect("should not be empty after we added something");
+            let memo: MemoPayload = AuthenticatedSenderMemo::new(
```

Comment:
An assert that tx_out_ref == tx_out might be a good idea, in case the order in which stuff gets added ever changes.